### PR TITLE
Actually set the image-to-index mapping in blobstore after it's generated and add a test that exercises that functionality

### PR DESCRIPTION
**Pull request by iain-macdonald**
**Related issues**: N/A

### CODE REVIEW

File: enterprise/server/sociartifactstore/sociartifactstore_test.go [hunk 1]
```diff
@@ -317,3 +348,22 @@ func (r *containerRegistry) Address() string {
 func (r *containerRegistry) ImageAddress(imageName string) string {
 	return fmt.Sprintf("%s:%d/%s", r.host, r.port, imageName)
 }
+
+type deduper struct {
+	mu sync.Mutex
+}
+
+func (d *deduper) Do(_ context.Context, _ string, work func() ([]byte, error)) ([]byte, error) {
+	d.mu.Lock()
```
Comment (vadimberezniker) [inline enterprise/server/sociartifactstore/sociartifactstore_test.go:357]: Why do you need a lock here?

File: enterprise/server/sociartifactstore/sociartifactstore_test.go [hunk 2]
```diff
@@ -317,3 +348,22 @@ func (r *containerRegistry) Address() string {
...
+	defer d.mu.Unlock()
+	return work()
+}
+
+func contains(s []string, e string) bool {
```
Comment (vadimberezniker) [inline enterprise/server/sociartifactstore/sociartifactstore_test.go:362]: you could use slices.Contains from golang.org/x/exp/slices, if you wanted to

File: server/real_environment/real_environment.go [hunk 3]
```diff
@@ -610,3 +611,10 @@ func (r *RealEnv) GetSociArtifactStoreServer() socipb.SociArtifactStoreServer {
 func (r *RealEnv) SetSociArtifactStoreServer(s socipb.SociArtifactStoreServer) {
 	r.sociArtifactStoreServer = s
 }
+
+func (r *RealEnv) GetSingleFlightDeduper() interfaces.SingleFlightDeduper {
+	return r.singleFlightDeduper
+}
+func (r *RealEnv) SetSingleFlightDeduper(d interfaces.SingleFlightDeduper) {
```
Comment (vadimberezniker) [inline server/real_environment/real_environment.go:618]: you are net setting the real deduper in env anywhere

File: enterprise/server/sociartifactstore/sociartifactstore_test.go [hunk 4]
```diff
@@ -290,24 +305,40 @@ func setup(t *testing.T) (*testenv.TestEnv, *SociArtifactStore, containerRegistr
...
 	go func() { _ = server.Serve(lis) }()
-	return r
+	return &registry
 }
 
 type containerRegistry struct {
 	host string
 	port int
+
+	// List of blobs which were requested using GET
```
Comment (vadimberezniker) [inline enterprise/server/sociartifactstore/sociartifactstore_test.go:340]: comment out of date?

**Review by vadimberezniker — CHANGES_REQUESTED** [review]

File: enterprise/server/sociartifactstore/sociartifactstore_test.go (hunk 4 above)
Comment (iain-macdonald) [reply enterprise/server/sociartifactstore/sociartifactstore_test.go]: Yup, fixed

File: enterprise/server/sociartifactstore/sociartifactstore_test.go (hunk 1 above)
Comment (iain-macdonald) [reply enterprise/server/sociartifactstore/sociartifactstore_test.go]: It's not needed, I just added it to be slightly more consistent with the deduper contract but am happy to just remove it instead.

File: enterprise/server/cmd/server/main.go [hunk 5]
```diff
@@ -302,5 +303,7 @@ func main() {
 		log.Fatalf("%v", err)
 	}
 
+	realEnv.SetSingleFlightDeduper(dsingleflight.New(realEnv.GetDefaultRedisClient()))
```
Comment (vadimberezniker) [inline enterprise/server/cmd/server/main.go:306]: redis is optional so you may want to make this conditional. we typically do this by adding a `Register(env)` function in the component itself.

File: enterprise/server/sociartifactstore/sociartifactstore.go [hunk 6]
```diff
@@ -105,7 +102,6 @@ func NewSociArtifactStore(env environment.Env) (error, *SociArtifactStore) {
 		cache:     env.GetCache(),
 		blobstore: env.GetBlobstore(),
 		env:       env,
```
Comment (vadimberezniker) [inline enterprise/server/sociartifactstore/sociartifactstore.go:104]: you'll want to skip setting up the artifact store if deduper is not available in the environment (if redis is not configured)

**Review by vadimberezniker — CHANGES_REQUESTED** [review]

File: enterprise/server/cmd/server/main.go (hunk 5 above)
Comment (iain-macdonald) [reply enterprise/server/cmd/server/main.go]: oh good call, fixed

File: enterprise/server/sociartifactstore/sociartifactstore.go (hunk 6 above)
Comment (iain-macdonald) [reply enterprise/server/sociartifactstore/sociartifactstore.go:112]: fixed

**Review by vadimberezniker — APPROVED** [review]

**Pull request merged**